App/Services/RpcController.cs

```diff
@@ -0,0 +1,248 @@
+    {
+        using var _ = await AcquireOperationLockNowAsync();
+        EnsureRpcConnected();
+
+        var credentials = _credentialManager.GetCredentials();
+        if (credentials.State != CredentialState.Valid)
+            throw new RpcOperationException("Cannot start VPN without valid credentials");
+
+        MutateState(state => { state.VpnLifecycle = VpnLifecycle.Starting; });
+
+        ServiceMessage reply;
+        try
+        {
+            reply = await _speaker!.SendRequestAwaitReply(new ClientMessage
+            {
+                Start = new StartRequest
+                {
+                    CoderUrl = credentials.CoderUrl,
+                    ApiToken = credentials.ApiToken,
+                },
+            }, ct);
+            if (reply.MsgCase != ServiceMessage.MsgOneofCase.Start)
+                throw new InvalidOperationException($"Unexpected reply message type: {reply.MsgCase}");
+        }
+        catch (Exception e)
+        {
+            MutateState(state => { state.VpnLifecycle = VpnLifecycle.Stopped; });
+            throw new RpcOperationException("Failed to send start command to service", e);
+        }
+
+        if (!reply.Start.Success)
+        {
+            MutateState(state => { state.VpnLifecycle = VpnLifecycle.Stopped; });
+            throw new VpnLifecycleException("Failed to start VPN",
+                new InvalidOperationException($"Service reported failure: {reply.Start.ErrorMessage}"));
+        }
+
+        MutateState(state => { state.VpnLifecycle = VpnLifecycle.Started; });
+    }
+
+    public async Task StopVpn(CancellationToken ct = default)
+    {
+        using var _ = await AcquireOperationLockNowAsync();
+        EnsureRpcConnected();
+
+        MutateState(state => { state.VpnLifecycle = VpnLifecycle.Stopping; });
+
+        ServiceMessage reply;
+        try
+        {
+            reply = await _speaker!.SendRequestAwaitReply(new ClientMessage
+            {
+                Stop = new StopRequest(),
+            }, ct);
+        }
+        catch (Exception e)
+        {
+            throw new RpcOperationException("Failed to send stop command to service", e);
+        }
+        finally
+        {
+            // Technically the state is unknown now.
+            MutateState(state => { state.VpnLifecycle = VpnLifecycle.Stopped; });
+        }
+
+        if (reply.MsgCase != ServiceMessage.MsgOneofCase.Stop)
+            throw new VpnLifecycleException("Failed to stop VPN",
+                new InvalidOperationException($"Unexpected reply message type: {reply.MsgCase}"));
+        if (!reply.Stop.Success)
+            throw new VpnLifecycleException("Failed to stop VPN",
+                new InvalidOperationException($"Service reported failure: {reply.Stop.ErrorMessage}"));
+    }
+
+    private void MutateState(Action<RpcModel> mutator)
+    {
+        using var _ = _stateLock.Lock();
+        mutator(_state);
+        StateChanged?.Invoke(this, _state.Clone());
```

Comment:
I think this will synchronously invoke the event handlers while holding the lock. Are we sure we want that? Seems like it could lead to deadlocks if the handler tries to read the state directly.